ta/pkcs11/src/persistent_token.c

```diff
@@ -538,6 +539,44 @@ void release_persistent_object_attributes(struct pkcs11_object *obj)
 	obj->attributes = NULL;
 }
 
+enum pkcs11_rc update_persistent_object_attributes(struct pkcs11_object *obj)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_ObjectHandle hdl = TEE_HANDLE_NULL;
+	uint32_t tee_obj_flags = TEE_DATA_FLAG_ACCESS_READ |
+				 TEE_DATA_FLAG_ACCESS_WRITE |
+				 TEE_DATA_FLAG_ACCESS_WRITE_META;
+	size_t size = 0;
+
+	if (obj->attributes)
+		TEE_Panic(0);
+
+	size = sizeof(struct obj_attrs) + obj->attributes->attrs_size;
+
+	res = TEE_OpenPersistentObject(TEE_STORAGE_PRIVATE,
+				       obj->uuid, sizeof(*obj->uuid),
+				       tee_obj_flags, &hdl);
+	if (res) {
+		EMSG("OpenPersistent failed %#"PRIx32, res);
+		return tee2pkcs_error(res);
+	}
+
+	res = TEE_TruncateObjectData(hdl, size);
```

Comment:
I have been thinking about the power cut endurance of the secure storage. Never really gone in the field deep enough yet but this got me thinking this once again.

I believe it would be better to:
- Open file
- Write data
- Truncate
- Close

This way file will always have correct data at the beginning (assuming that write at least goes to the storage) even if the truncate is lost.

Now if the truncate is made first and that succeeds and then power is lost then we have object that is possibly incomplete as new data is not yet written.

If power cut is in within the write operation then we have lost the game and that object is broken which is bad.

In order to tackle that I suppose we would need to create a new persistent object and then change that in place of old object and then delete the old object.

Reply:
You got something here.
I think open/write/truncate/close looks the right sequence. And a sanity check/clean at database open may be needed to ensure perssitent object are clean. There would also address case where we create a new persistent object, got the object written in persistent object but its reference not written in the persistent dabase object because of some power loss, reboot or like.
I think you should address these case in a specific P-R.

Reply:
Sequence you mentioned seems logical @vesajaaskelainen. I will correct it for this PR. But we have other instances in code where same sequence is followed. As @etienne-lms mentioned, we can address that as a separate PR.